snuba/utils/streams/abstract.py

```diff
@@ -0,0 +1,137 @@
+    value: TValue
+
+
+class ConsumerError(Exception):
+    """
+    Base class for exceptions that are raised during consumption.
+
+    Subclasses may extend this class to disambiguate errors that are specific
+    to their implementation.
+    """
+
+
+class EndOfStream(ConsumerError, Generic[TStream]):
+    """
+    Raised when there are no more messages to consume from the stream.
+    """
+
+    def __init__(self, stream: TStream):
+        self.stream = stream
+
+
+class Consumer(ABC, Generic[TStream, TOffset, TValue]):
+    """
+    An abstract class that provides methods to consume messages from a
+    multiplexed collection of streams.
+
+    This interface is heavily "inspired" by the Confluent Kafka Consumer
+    implementation, but only exposes a limited set of the available methods
+    and several method signatures and/or return values differ -- these
+    changes are intended to improve the ease of use of the consumer, as well
+    as make development and testing more straightforward. Most code that uses
+    the Confluent driver can be modified to use this interface with minimal
+    changes, but it is not intended to be a drop-in replacement for the
+    Confluent implementation.
+
+    All methods are blocking unless otherise noted in the documentation for
+    that method.
+    """
+
+    @abstractmethod
+    def subscribe(
+        self,
+        topics: Sequence[str],
+        on_assign: Optional[Callable[[Sequence[TStream]], None]] = None,
+        on_revoke: Optional[Callable[[Sequence[TStream]], None]] = None,
+    ) -> None:
+        """
+        Subscribe to topic streams. This replaces a previous subscription.
+        This method does not block. The subscription may not be fulfilled
+        immediately: instead, the ``on_assign`` and ``on_revoke`` callbacks
+        are called when the subscription state changes with the updated
+        assignment for this consumer.
+
+        Raises a ``RuntimeError`` if called on a closed consumer.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def poll(
+        self, timeout: Optional[float] = None
+    ) -> Optional[Message[TStream, TOffset, TValue]]:
+        """
+        Return the next message available to be consumed, if one is
+        available. If no message is available, this method will block up to
+        the ``timeout`` value before returning ``None``. A timeout of
+        ``0.0`` represents "do not block", while a timeout of ``None``
+        represents "block until a message is available (or forever)".
+
+        Calling this method may also invoke subscription state change
+        callbacks.
+
+        If the backend supports it, this method may raise a ``EndOfStream``
+        error if there are no messages available to be consumed (and the
+        backend is configured to do so.) This may also raise
+        ``ConsumerError`` or one of it's subclasses, the meaning of which are
+        implementation-dependent.
```

Comment:
> While the confluent api represent the EOF as an error, I don't think it is. Polling from a stream and not finding anything seems a pretty standard condition to me. Have you considered alternatives to throwing an exception to signal that state ?

That's true. The only other alternative I can think of would be to take `ConsumerError` out of the `EndOfStream` hierarchy and return `Union[Message[TStream, TOffset, TValue]], EndOfStream[TStream], None]` from this method instead.

Advantages:

- The end of the stream is not necessarily an error.
- It's likely faster to avoid the exception machinery in the runtime.

Disadvantages:

- `EndOfStream` will be part of the type signature even if it is disabled in the consumer configuration, so you'll still have to handle it to satisfy the type checker even if it will never be returned from this method.

> Also, in the Kafka implementation, things are more tricky, since you get an EOF per partition, so the consumer may still have plenty of valid messages to consume and still throw because one partition is at the end of the stream.

I think that's just a matter of updating the documentation, this could be clearer about the fact that streams are multiplexed beyond just the mention in the class docstring here. My assumption was that anybody who needed the EOF (which to be clear, is nobody using this right now) would need to `try`/`except` this the same way you have to check `message.error()` with the Confluent consumer.

Reply:
Sounds good, I'll update the documentation. The more that I've thought about it, the stronger I feel that the EOF is a special case for most consumers, so I think overall this is the better solution for most consumers.